d3-scripts/src/d3_scripts/get_claim_tree.py

```diff
@@ -0,0 +1,78 @@
+from typing import Sequence, List, Mapping, Any
+
+BehaviourJson = Mapping[str, Any]
+BehaviourJsons = Sequence[BehaviourJson]
+
+
+def get_claim_tree(claim: dict, behaviour_jsons: BehaviourJsons):
+    """
+    Resolve claim inheritance tree, as well as validate parents of claim exist and that parents doesn't include
+    claim itself
+
+    Args:
+        claim: The D3 claim to get behaviour claim tree for (dict)
+        behaviour_jsons: The array of behaviours (array of dict)
+
+    Returns:
+        Claim inheritance tree. First element is claim, next element is parents of claim,
+        next element is parents of parents of claim and so on. Final element should be an empty list with second to
+        last element being final parents in inheritance tree. (array of array of dict)
```

Comment:
@aloisklink I had a look at anytree and it doesn't work for our use-case, due to multiple parents, the data structure we need to represent the inheritance is a Directed Acyclic Graph, we could use [networkx](https://networkx.org/) like so: https://mungingdata.com/python/dag-directed-acyclic-graph-networkx/. Do you think it's worth using this data structure instead of an array of arrays?

Reply:
> due to multiple parents

Ahhhh, would it be possible to invert the tree? So instead of having multiple parents, you have multiple children? (of course, all the arrows in the diagram will be in the wrong direction, and maybe it's harder to build a tree in reverse?)

Just a side note, the last time I did tree-like structures was in high school, so it was years ago for me :)

My opinion is yeah, it's probably worth using a data structure.
https://networkx.org/documentation/latest/tutorial.html#directed-graphs

Ideally, instead of using your custom iterator functions, calling `node.parents()`/ `node.children()`/`digraph.successors()` to do it automatically would be ideal.
Plus that basically makes the code a lot more self-documenting.

Also, having a nice 2D image of the graph structure would make debugging a bunch easier :)

But it's up to you and @cbrafter on how urgent it is. If you feel like it's too much effort for too little gain, feel free to merge without making this change.

Reply:
Btw, if you do make a graph/tree, you may need to create a hashable D4 claim.

Might be easiest to do it with https://docs.python.org/3/library/dataclasses.html

```python3
from dataclasses import dataclass

@dataclass(eq=True, frozen=True) # setting eq/frozen means the class will be __hash__()-able
class D3Claim:
    """Class for keeping track of a D3 Claim."""
    name: str
    parents: List[D3Claim]
```

Reply:
It does look like `anytree` will work for you https://anytree.readthedocs.io/en/latest/

It looks like the iteration method it uses is `node -> children (aka d3 parents)`. Which I believe is what we want? If you have a D3 claim, you only care about its parents, not its children.

The best thing is, it looks pretty easy to make a custom class called `D3 Claim`.

You just need to do something like https://anytree.readthedocs.io/en/latest/api/anytree.node.html#anytree.node.nodemixin.NodeMixin

Pseudo-code below:

```python
from anytree import NodeMixin

class D3Claim(NodeMixin):
  def _init_(uuid, claim_json, children):
    super(MyClass, self).__init__()
    self.uuid = uuid
    self.claim_json = claim_json
    if children:
        self.children = children
  @cache
  @static_method
  def from_uuid(uuid):
    claim_json = load_claim(uuid)
    d3_parent_nodes = [D3Claim(d3_parent_uuid for d3_parent_uuid of self.claim["parents"]]
    # in "anytree.children" are equivalent to d3 parents
    return D3Claim(
      uuid,
      claim_json, # could remove this bit if it uses a lot of RAM
      children=d3_parent_nodes,
    )
```